src/kcas.ml

```diff
@@ -8,181 +8,167 @@ module type Backoff = Backoff.S
-
-let st_eq s s' =
-  match (s, s') with
-  | WORD x, WORD x' -> x == x'
-  | RDCSS_DESC r, RDCSS_DESC r' -> r.id_rdcss == r'.id_rdcss
-  | CASN_DESC c, CASN_DESC c' -> c.id_casn == c'.id_casn
-  | _ -> false
-
-let commit (CAS (r, expect, update)) =
-  let s = Atomic.get r.content in
-  st_eq s expect && Atomic.compare_and_set r.content s update
-
-let cas r e u = commit (mk_cas r e u)
-
-let set r n = Atomic.set r.content (WORD n)
-
+let is_on_ref (T (r1, _, _)) r2 = equal r1 r2
+let mk_cas a old_value new_value = T (a, old_value, new_value)
 let get_id r = r.id
 
-let rec rdcss rd =
-  if commit (CAS (rd.a2, rd.o2, RDCSS_DESC rd)) then (
-    ignore @@ complete rd ;
-    rd.o2 )
-  else
-    let r = Atomic.get rd.a2.content in
-    match r with
-    | RDCSS_DESC rd' ->
-        ignore @@ complete rd' ;
-        rdcss rd
-    | _ -> if st_eq r rd.o2 then rdcss rd else r
-
-and complete rd =
-  if st_eq (Atomic.get rd.a1.content) rd.o1 then
-    commit (CAS (rd.a2, RDCSS_DESC rd, rd.n2))
-  else commit (CAS (rd.a2, RDCSS_DESC rd, rd.o2))
-
-let rec rdcss_read a =
-  let r = Atomic.get a in
-  match r with
-  | RDCSS_DESC rd ->
-      ignore @@ complete rd ;
-      rdcss_read a
-  | _ -> r
-
-let rec casn_proceed c =
-  let rec phase1 curr_c_l curr_st out =
-    match curr_c_l with
-    | CAS (a, o, n) :: curr_c_t_tail when curr_st = SUCCEEDED -> (
-        let s = rdcss (mk_rdcss c.st (WORD UNDECIDED) a o (CASN_DESC c)) in
-        match s with
-        | CASN_DESC c' ->
-            if c.id_casn != c'.id_casn then (
-              ignore @@ casn_proceed c' ;
-              phase1 curr_c_l curr_st out )
-            else phase1 curr_c_t_tail curr_st (CAS (a, o, n) :: out)
-        | _ ->
-            if st_eq s o then
-              phase1 curr_c_t_tail curr_st (CAS (a, o, n) :: out)
-            else phase1 curr_c_t_tail FAILED (CAS (a, o, n) :: out) )
-    | _ ->
-        ignore @@ commit (CAS (c.st, WORD UNDECIDED, WORD curr_st)) ;
-        out
-  in
-  let rec phase2 curr_c_l succ =
-    match curr_c_l with
-    | CAS (a, o, n) :: curr_c_l_tail -> (
-      match Atomic.get succ with
-      | WORD SUCCEEDED ->
-          ignore @@ commit (CAS (a, CASN_DESC c, n)) ;
-          phase2 curr_c_l_tail succ
-      | _ ->
-          ignore @@ commit (CAS (a, CASN_DESC c, o)) ;
-          phase2 curr_c_l_tail succ )
-    | _ -> Atomic.get succ = WORD SUCCEEDED
-  in
-  match Atomic.get c.st.content with
-  | WORD UNDECIDED -> phase2 (phase1 c.c_l SUCCEEDED []) c.st.content
-  | _ -> phase2 c.c_l c.st.content
-
-let rec get a =
-  let r = rdcss_read a.content in
-  match r with
-  | CASN_DESC c ->
-      ignore @@ casn_proceed c ;
-      get a
-  | WORD out -> out
-  | _ -> assert false
-
-let kCAS c_l =
-  match c_l with
+let set atom value =
+  Atomic.set atom.state { before = value; after = value; casn = casn_after }
+
+let rec release_after = function
+  | [] -> true
+  | CAS (_, cas) :: cass ->
+      cas.casn <- casn_after;
```

Comment:
The status has already been updated on line 65, do we need this one (and the one on line 52)?

Reply:
See the comment on line 56.  This reduces space (by freeing an atomic object) at the expense of a write barrier.

It is not entirely clear to me whether freeing the space is worth it.

I guess it boils down to usage patterns.  Speculation: If you constantly keep updating most kcas refs, then the freeing the space is probably not worth it.  If, OTOH, you have a lot of kcas refs and only update a few of those at any given moment, then it might be relatively more beneficial to free the space.